### DISCUSSION TITLE
zero-knowledge servers and subscriptions - how does that work?

### DISCUSSION THREAD

**Discussion by rozek** [logux/logux#126] [Q&A]
Good morning!

Sorry for my silly questions - but I'm currently investigating Logux and do not yet fully understand it.

My current question is: we have actions and subscriptions - and we have "zero knowledge" servers  with opaque action logs but without the actual data. How can clients then subscribe to any channels if those servers do not have the knowledge set them up?

My first thought was that subscriptions would be transparently passed on to clients - but clients cannot set-up any channels.

Thus: how do these two concepts (subscriptions and "zero knowledge") go together?

**Comment by ai** [comment]
> How can clients then subscribe to any channels if those servers do not have the knowledge set them up?

There are 2 ways.

1. If you do not need to keep old data, you just define [`server.channel()`](https://logux.org/node-api/#server-channel) (with empty `load`) and use `client.log.add({ type: 'logux/subscribe', … })`. Then server will resend actions to all subscribed clients without saving them to the server. But when new client connects it will need to wait new data from other clients (it could send some special action to ask other clients to send data).
2. You can use [end-to-end encryption](https://logux.org/web-api/#globals-encryptactions) and save encrypted data to server’s database (and load it in `server.channel` → `load` to send back to new client on subscription).

**Reply by rozek** [reply]
well, I like the idea of having servers which are just "message brokers" - such servers are lightweight and one may (perhaps) switch between such brokers for the sake of load balancing. These brokers, however, would not persist any data (that's "zero knowledge", not suitable for option 2)

The "special action" (let's call it "logux/load") would then have to be broadcasted in the "ChannelLoader"? Would that loader itself then have to respond with an empty action list and wait for any client to provide the actual data? And, would the broker then be smart enough to relay only the first such response ("meta" should be left unencrypted to my understanding)?

(Subscription actions would be "ephemeral" and not be kept in the log)

**Reply by ai** [reply]
> (let's call it "logux/load") 

`logux/*` prefix is only for action by Logux team :D. You must use `data-load`.

> The "special action" (let's call it "logux/load") would then have to be broadcasted in the "ChannelLoader"? Would that loader itself then have to respond with an empty action list and wait for any client to provide the actual data?

Yes, server will just resend `data-load` to all client and so they will send data back to the new client.

> And, would the broker then be smart enough to relay only the first such response ("meta" should be left unencrypted to my understanding)?

No, you need to handle it in `server.type()`. Also you need to create some random `requestId` to `load-data` and `load-data/repsonse`. Then you need some `Map` on the client to map `requestId` to `clientId`.

**Reply by ai** [reply]
> (Subscription actions would be "ephemeral" and not be kept in the log)

Yes. All server log is deleted by default (you need manually set “reasons to live” to actions into `meta.reasons`)

**Reply by rozek** [reply]
good ideas, thanks!

**Comment by ai** [comment]
> but clients cannot set-up any channels

To subscribe to any channel you need to send `logux/subscribe` action:

```js
client.log.add({ type: 'logux/subscribe', channel: 'channelName' })
```
